### PR TITLE
Update DSpace with Jakarta EE Support (includes Tomcat 10+)

### LINKED ISSUE #8173: Backend (Hal Browser) doesn't work properly in Tomcat 10
**Describe the bug**
Originally reported in #8114

When Tomcat 10 is installed, the HAL Browser refuses to load its static resources/styles.  This means the DSpace 7.x backend currently only works on Tomcat 9.

_NOTE: It's unclear at this time if the bug is specific to the HAL Browser.  If so, it might be that solving #3017 will also fix this issue._

**To Reproduce**
1. Attempt to run the DSpace backend on Tomcat 10
2. Result will be similar to #8114 and as reported in https://groups.google.com/g/dspace-tech/c/VbPY-N3IDMY/m/re__nCKXBwAJ

**Expected behavior**
Tomcat 10 should work.

**Related work**
#8114

### LINKED ISSUE #8713: DSpace 7 doesn't support Tomcat 10 (requires Tomcat 9)
**Describe the bug**
DSpace 7 doesn't support Tomcat 10, because Tomcat 10 is not backwards compatible with Tomcat 9.

The most significant change is that Tomcat 10 now uses Jakarta Servlet 5.0 (and 10.1 uses Jakarta Servlet 6.0).  This is incompatible with older versions of Jakarta Servlet because the class packages have changed from `javax.*` to `jakarta.*`).  Tomcat 9 is the last release to support `javax.*` packages.

DSpace 7 currently uses `javax.*` packages in its backend code.  It's not possible to support both `javax.*` and `jakarta.*` packages at the same time.  So, the soonest DSpace can support Tomcat 10 would be with version 8.0.

**To Reproduce**
* Attempt to deploy DSpace 7 on Tomcat 10.  You'll see this error in the logs:
```
10-Mar-2023 20:54:57.824 SEVERE [main] org.apache.catalina.core.StandardContext.listenerStart Error configuring applicat
ion listener of class [org.dspace.app.util.DSpaceWebappListener]
        java.lang.NoClassDefFoundError: javax/servlet/ServletContextListener
                at java.base/java.lang.ClassLoader.defineClass1(Native Method)
                at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
                at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174)
                at org.apache.catalina.loader.WebappClassLoaderBase.findClassInternal(WebappClassLoaderBase.java:2487)
                at org.apache.catalina.loader.WebappClassLoaderBase.findClass(WebappClassLoaderBase.java:865)
                at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1401)
                at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1245)
                at org.apache.catalina.core.DefaultInstanceManager.loadClass(DefaultInstanceManager.java:490)
                at org.apache.catalina.core.DefaultInstanceManager.loadClassMaybePrivileged(DefaultInstanceManager.java:472)
``` 
This error is the result of attempting to use `javax.*` packages on Tomcat 10, which only supports `jakarta.*` packages.

This error also results in #8173.  It appears to simply be a side effect of the incompatibility with Tomcat 10.

**Expected behavior**
DSpace should eventually support Tomcat 10.  However, that is unlikely to occur until version 8.0, because it requires code changes which will make DSpace incompatible with Tomcat 9.

**Related work**
Related to #8173

### LINKED ISSUE #9119: Upgrade Flyway to v9
DSpace is currently using Flyway 8.4.4, but the latest version is [9.22.2](https://mvnrepository.com/artifact/org.flywaydb/flyway-core/9.22.2). Officially, Flyway 8 only supports PostgreSQL up to version 14, while Flyway 9 includes support for PostgreSQL 15.

Since DSpace lists 15 as supported PostgreSQL version it should IMHO also use the respectable Flyway version.

Flyway 8 *seems* to run without problems with PSQL 15 though. Flyway 9 is a fairly straightforward upgrade, the only relevant API change looks to be the removal of `flywayConfiguration.ignoreIgnoredMigrations(true);` in favor of `flywayConfiguration.ignoreMigrationPatterns(String ...)`

Perhaps it would be sensible to wait until Flyway releases a version which is compatible with PSQL 16 before the upgrade

### LINKED ISSUE #9209: Upgrade to Spring Boot 3.x
The final release of Spring Boot 2.7.x has just been released:  https://spring.io/blog/2023/11/23/spring-boot-2-7-18-available-now

We need to upgrade therefore to Spring Boot 3.x per the migration guide: https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.0-Migration-Guide

Major (known) changes in Spring Boot 3:
* JDK 17 or later is required
* Spring Security 6 is required: https://docs.spring.io/spring-security/reference/6.0/migration/index.html
* Spring Framework 6 is required: https://github.com/spring-projects/spring-framework/wiki/Upgrading-to-Spring-Framework-6.x

This is likely a significant upgrade which cannot be easily backported to DSpace 7.x.

### PR DESCRIPTION
**Pull request by tdonohue**
## References
* Fixes #8713
* Fixes #9209
* Fixes #8173
* Fixes #9119 
* **REST Contract for new `GET /api/security/csrf` endpoint:** https://github.com/DSpace/RestContract/pull/259

## Description
Migration of DSpace codebase from Java Enterprise Edition (Jave EE) to [Jakarta Enterprise Edition (Jakarta EE)](https://en.wikipedia.org/wiki/Jakarta_EE).

**Breaking changes:**
* Java 17+ is now required. (Java 11 and below no longer supported)
    * NOTE: DSpace 7 also runs fine on Java 17. So all developers should move to Java 17.
* Tomcat 10+ is now required. (Tomcat 9 is no longer supported)
* Most `javax.*` dependencies migrated to `jakarta.*` equivalents

**Necessary upgrades:**
* Spring Boot (v2.7.18 -> v3.2.4)
* Spring (v5.3.32 -> v6.1.5)
* Spring Security (v5.7.11 -> v6.2.3)
* Hibernate (v5.6.15.Final -> v6.4.4.Final)
* FlywayDB (v8.5.13 -> v10.10.0)
* Jena (v2 -> v4)
* Jersey (v2 -> v3)
* SLF4j (v1 -> v2) (we don't use SLF4j directly, so this is just for dependency convergence purposes)

**Other upgrades:**
While performing the upgrades above, I also took the liberty of adding a few more upgrades into our main POM
* Log4j (v2.22.1 -> v2.23.1)
* PDFBox (v2.0.30 -> v2.0.31)
* Tika (v2.5.0 -> v2.9.1)
* Nimbus JOSE+JWT (v7.9 -> v9.37.3)
* All maven-*-plugins have been updated to latest version

**Other major changes:**
* Migrated SWORDv2 to using the version distributed by [GDCC](https://www.gdcc.io/): https://github.com/gdcc/sword2-server
* Migrated to using the `orcid-model` API distributed by [OICR](https://oicr.on.ca/) via https://artifacts.oicr.on.ca/artifactory/collab-release/ . OICR has built their own version of the `orcid-model` API which is compatible with Jakarta EE.
    * This is a temporary change until ORCID distributes an official version of this API which is compatible with Jakarta EE. See https://github.com/ORCID/orcid-model/issues/50
    * Alternatively, we also could consider regenerating/redistributing our own `orcid-model` API by reopening this module: https://github.com/DSpace/orcid-jaxb-api
* Handle Server's `cnri-servlet-container` and its required Jetty dependencies are all **runtime** dependencies now.  Previously these were compile-time, but they are only necessary to startup the handle server. This was also necessary because Jetty is not fully compatible with Jakarta EE.  (I tested these changes locally and the Handle Server appears to still work fine)

## Instructions for Reviewers
* Perform a code review
* Ensure application is working with this updated backend.  All features should work.

### CODE REVIEW

**Comment by mwoodiupui** [comment]
Might this be a good opportunity to remove DSpace's direct dependency on SLF4J?  Log4Jv2 should be able to replace everything we are doing with it.  I've been meaning to work up a patch to do that, and could contribute such work to the Jakarta effort.

There will still be indirect dependencies on SLF4J that we may need to coordinate.

**Comment by tdonohue** [comment]
@mwoodiupui : I'd be perfectly OK with removing SLF4J if we can do so.  It might even be possible to do on `main` **before** this larger Jakarta refactor takes place (in which case I can remove any SLF4J changes from this PR).  If you think that'd work, please develop a PR against `main` and assign me as the reviewer.  I'll get it reviewed/merged quickly and then rebase this PR on top of those changes.

So, If you are willing, I'd support that.  Otherwise, I'm not sure whether I can add that *myself* into this PR... there are too many other things that need fixing. (Currently this PR is non-functional.  It's only a partial update & is waiting for me to get back to it again.)

**Comment by jeffmorin** [comment]
I think this PR is a good opportunity to validate the updated code base against Java 21, which is after all the latest LTS version of Java. In the worst case, if something doesn't work properly, it will always be possible to switch back to Java 17.

**Comment by tdonohue** [comment]
@mwoodiupui and @abollini :  I've unfortunately run into some upgrade issues with the migration from Spring Security v5 -> v6 as much of our existing authentication & authorization & CSRF code required refactoring to align with the new version.  These refactors were time consuming & are **still not 100% complete**.  The code is in a better sate, but parts of authentication/authorization are not working properly (especially the CSRF-related code)  

So, while you could start some early review on this PR... it is not yet fully functional.  Tests are mostly passing but there are still ITs to cleanup.  It's become much more time consuming than I even expected.  I'll be revisiting this next week (when I return from vacation on Monday, March 18) in the hopes of finally getting this working.

In the meantime, if either of you find a spare moment you are also welcome to send PRs to my branch.  No worries though if that's not possible on your end.

**Comment by tdonohue** [comment]
This PR now mostly "works" in that the backend can be run successfully, HAL browser works and the UI mostly "works".  Therefore, I'm taking this PR out of draft stage.

That said, this PR is still **not** fully complete.  As mentioned in today's DevMtg, there are still failing ITs which I'm working on fixing. Namely ITs related to the following:

1. CSRF Tokens are *not* fully working. This causes failures in several tests in `AuthenticationRestControllerIT`. 
    * This appears to be a result to major changes to CSRF tokens in Spring Security v6 / Spring Boot v3.
2. Some `PATCH` commands which use the "move" action are not working. This causes failures in `PatchMetadataIT`.  
    * **I'm not yet certain of the root cause.**  No errors appear on the backend.  Instead the "move" action return success (200 OK) while **not** reordering the fields (i.e. the move essentially seems to do nothing?).

**Comment by mwoodiupui** [comment]
If ORCiD code is still an issue, we may need to look at the https://github.com/apache/tomcat-jakartaee-migration tool which, I believe, will do bytecode surgery.

**Comment by tdonohue** [comment]
@mwoodiupui : I've found a temporarily replacement for the official `orcid-model` API.  I added a dependency on the `orcid-model` API distributed by https://artifacts.oicr.on.ca/artifactory/collab-release/ (from https://oicr.on.ca/).  They have built their own version of the `orcid-model` API which is compatible with Jakarta EE. 

We also could consider doing the same by simply recreating/redistributing this model again: https://github.com/DSpace/orcid-jaxb-api   But, for now, the "quick win" was to use the version from OICR, as it appears to work (at least it passes all our ORCID tests).

I'll update the description above with these details.

**Comment by tdonohue** [comment]
UPDATE on CSRF code:
* As of commit cda2387, the CSRF tokens are **working** if you run the backend.  This means you can successfully login via the UI and also perform all main activities (edit, delete, create). 
* HOWEVER, this commit has caused **nearly all Integration Tests** to fail, because the way we do authentication in ITs doesn't seem compatible with these new CSRF tokens.  (You can see the prior commit 0da79b7 shows most ITs succeeding.)

So, the **good news** is this PR is now testable with the UI.  It's mostly working.

The **bad news** is that all the ITs that use authentication are no longer working.  I'll have to dig into why on Monday.

**Comment by tdonohue** [comment]
As of yesterday, the CSRF token issues are now **fixed** and all authentication/CSRF related tests are passing.  The only remaining failures are in these tests:
* `PatchMetadataIT` - several tests related to "move" operation
* `ComColAdminFeatureIT` - This one is failing in GitHub CI but succeeding locally. I'm digging into it.
* `NotifyServiceRestRepositoryIT` - One test in this class is failing. I'm digging into it.

@mwoodiupui & @abollini : This means this PR is **ready for testing**.  I'll get these final ITs cleaned up, but the overall backend should be working.   Obviously, we won't merge this until a solution for these final ITs is found, but I'd encourage you both to start testing/reviewing this PR while I finish fixing these last few ITs.

**Comment by tdonohue** [comment]
@abollini and @mwoodiupui : As of today, this PR is complete.  **All tests pass.**  This is also working for me locally.  Please review/test this as soon as you can so that we can merge this quickly (and allow for additional bug fix PRs to be merged).

The only remaining build failure is a Docker build issue which _should be ignored_.  It only occurring because this PR changes our build requirements to Java 17+ and the latest DSpace Docker images on DockerHub all use Java 11. So, the build failure is because this PR is attempting to use the Java 11 images from DockerHub to compile this Java 17 code. The failure will go away as soon as this PR is merged (as the latest images on DockerHub will then all be updated to use Java 17).

**Review by abollini — APPROVED** [review]
thanks @tdonohue for all your work on that. Code changes looks good to me and I have tested that all the basic functionalities work properly on my side. I'm also happy to say that with the update to java 17 it is now possible to run dspace in the latest version of Eclipse (just few additional changes are needed, I'm created a PR for that)

File: dspace-api/src/main/java/org/dspace/authorize/AuthorizeServiceImpl.java
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/authorize/AuthorizeServiceImpl.java:700]: This was a situation where Hibernate 6 is much less "forgiving" of null values in queries.  This block of code runs `findByTypeGroupAction()`, but if the `group` is null this throws errors in Hibernate.  That's why this check had to be modified to first verify that the `group` is not null (as otherwise tests which only passed in an EPerson would fail with Hibernate errors).

So, I think this change is correct and isn't an indication of a bug. The code in the "else if" seems to suggest that a `group` is required based on this comment:
> // if an identical policy (same Action and same Group) is already in place modify it

File: dspace-api/src/main/java/org/dspace/authorize/dao/impl/ResourcePolicyDAOImpl.java
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/authorize/dao/impl/ResourcePolicyDAOImpl.java:163]: I was attempting to model the existing behavior here, as this is another example of code where Hibernate 6 is much less forgiving about null values.  In Hibernate 5, if the Groups & Eperson were empty, this method would *still run* but would always return no results.  In Hibernate 6, if these objects are null/empty, then the method was throwing a Hibernate Syntax error when the query was run.  That's why I had to modify this method to ensure we are not passing null or empty values to Hibernate.  

So, this code is attempting to ensure we keep the same behavior as before... which is why an empty set of results is returned instead of an error.

I'm not against changing the behavior to throw an exception, but that would likely require additional code changes to methods which call this one.  I think that's out of scope for this PR, but it's something we could do later on if we felt it was necessary.

File: dspace-api/src/main/java/org/dspace/authorize/AuthorizeServiceImpl.java
Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/authorize/AuthorizeServiceImpl.java:700]: Would've preferred Hibernate to call it an illegal-argument error but I like the new behavior better.  And we should not even ask Hibernate if we are asking for nothing.

File: dspace-api/src/main/java/org/dspace/authorize/dao/impl/ResourcePolicyDAOImpl.java
Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/authorize/dao/impl/ResourcePolicyDAOImpl.java:163]: Again we should not clank the database when we *know* there will be no result; we should just return no result.  I wonder whether we should log a warning, but will not argue for it at this time.

We should probably follow up with a careful examination of our database code to comb out such places where we are lazy and lean on the DBMS to tell us that we asked for nothing.  AFTER this patch goes in.

I agree with the change in Hibernate.  Java `null` does not necessarily mean SQL `IS NULL`.  If you are searching for rows with null column values you should be required to explicitly say so.  The old behavior was probably the source of many mysterious errors.

Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/authorize/dao/impl/ResourcePolicyDAOImpl.java:163]: Yes, I agree that I actually *prefer* the new Hibernate behavior.  It forces us to filter out the null values *before* using Hibernate to query anything.  But, it did require some code cleanup in this PR to add those extra null/empty checks. I **think** I caught them all...but I agree we should keep an eye on our database queries better after this is merged.

@mwoodiupui : Do you want a chance to still check this PR further, or is it good to merge?  (I'd like to merge today/tomorrow if at all possible)

**Comment by tdonohue** [comment]
Merging with `1 Approval` as this needs to get added prior to Testathon & early enough to get some other fixes applied to `main`.  This is working for me as well, and @abollini has approved.

**Comment by tdonohue** [comment]
From this PR, I've ported the new `/api/security/csrf` endpoint to 7.x in #9599

**Pull request merged**
